hermes/probe/checknil/check_nil.go

```diff
@@ -0,0 +1,332 @@
+			// If two -> verify whih one reflects the current state of the system and delete the other.
+			// If two and only differ by intent, prefer the one with intent.
+			obj, err := objIter.Next()
+			if err == iterator.Done {
+				return probe.NewProbeError(metrics.FileMissing, fmt.Errorf("checkNilFileExists(): unable to find nil file: %w", err))
+			}
+			if err != nil {
+				return probe.NewProbeError(metrics.APICallFailed, fmt.Errorf("checkNilFileExists(): unable to list files: %w", err))
+			}
+			file.name = obj.Name
+			return probe.NewProbeError(metrics.Success, nil)
+		}
+	})
+
+	logger.Infof("checkNilFileExists(): nil file located on target storage system.")
+	return file, err
+}
+
+// checkNilFileExists checks that the nil file exists on the target system.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- nilFile: a nilFile struct with the file name recorded
+//	- err: error exit statuses are well defined in metrics.go
+func readNilFile(ctx context.Context, file *nilFile, bucket string, target *probe.Target, client stiface.Client, logger *logger.Logger) (*nilFile, *probe.ProbeError) {
+	fileObject := client.Bucket(bucket).Object(file.name)
+	var reader stiface.Reader
+
+	err := RecordAPILatency(target.LatencyMetrics.APICallLatency, metrics.APIGetFile, func() *probe.ProbeError {
+		var err error
+		reader, err = fileObject.NewReader(ctx)
+		if err != nil {
+			return probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("readNilFile(): unable to read nil file: %w", err))
+
+		}
+		return probe.NewProbeError(metrics.Success, nil)
+	})
+	defer reader.Close()
+
+	if err.Status != metrics.Success {
+		return nil, err
+	}
+
+	logger.Infof("readNilFile(): nil file retrieved from target storage system.")
+
+	var readErr error
+	file.contents, readErr = ioutil.ReadAll(reader)
+	if readErr != nil {
+		return nil, probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("readNilFile(): unable to read nil file: %w", readErr))
+	}
+
+	logger.Infof("readNilFile(): nil file read complete.")
+
+	return file, probe.NewProbeError(metrics.Success, nil)
+}
+
+// verifyNilFile verifies that the contents of the nil file match its checksum
+// and that the journal matches the existing in-memory proto.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- nilFile: a nilFile struct with the file name and contents recorded
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func verifyNilFile(ctx context.Context, file *nilFile, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
```

Comment:
This should be a method on `nilFile` as it modifies its contents.

Reply:
I have refactored this into 3 functions: one that modifies the contents of the journalFile. and two that perform operations on the journalFile but do not modify its contents.